plugins/base/src/main/kotlin/transformers/documentables/DefaultDocumentableMerger.kt

```diff
@@ -1,36 +1,78 @@
+import org.jetbrains.kotlin.utils.addToStdlib.safeAs
 
 internal object DefaultDocumentableMerger : DocumentableMerger {
+    var dependencyInfo: Map<DokkaConfiguration.DokkaSourceSet, List<DokkaConfiguration.DokkaSourceSet>>? = null //Sorry for that hack
 
     override fun invoke(modules: Collection<DModule>, context: DokkaContext): DModule {
+        this.dependencyInfo = context.getDependencyInfo()
 
         val projectName =
             modules.fold(modules.first().name) { acc, module -> acc.commonPrefixWith(module.name) }
                 .takeIf { it.isNotEmpty() }
                 ?: "project"
 
-        return modules.reduce { left, right ->
+        return topologicalSort(modules).reduce { left, right ->
             val list = listOf(left, right)
             DModule(
                 name = projectName,
                 packages = merge(
                     list.flatMap { it.packages },
                     DPackage::mergeWith
                 ),
-                documentation = list.map { it.documentation }.flatMap { it.entries }.associate { (k,v) -> k to v },
+                documentation = list.map { it.documentation }.flatMap { it.entries }.associate { (k, v) -> k to v },
                 expectPresentInSet = list.firstNotNullResult { it.expectPresentInSet },
                 sourceSets = list.flatMap { it.sourceSets }.toSet()
             ).mergeExtras(left, right)
         }
     }
 }
 
+private fun topologicalSort(modules: Collection<DModule>) : List<DModule> {
+    val modulesMap = modules.map { it.sourceSets.single().sourceSetID to it }.toMap()
+    val graph = modules.flatMap { module ->
+        module.sourceSets.single().
+        dependentSourceSets
+            .map {
+                    srcset -> modulesMap[srcset]!! to module
+            }
+    }
+        .groupBy { it.first }.entries
+        .map { it.key to it.value.map { it.second } }
+        .toMap()
+    //this returns representation of graph where directed edges are leading from module to modules that depend on it
+    val visited = modules.map { it to false }.toMap().toMutableMap()
+    val topologicalSortedList : MutableList<DModule> = mutableListOf()
+
+    fun dfs(module: DModule) {
+        visited[module] = true
+        graph[module]?.forEach { if (!visited[it]!!) dfs(it) }
+        topologicalSortedList.add(0, module)
+    }
+    modules.forEach { if (!visited[it]!!) dfs(it) }
+
+    return topologicalSortedList
+}
+
+private fun DokkaContext.getDependencyInfo()
+        : Map<DokkaConfiguration.DokkaSourceSet, List<DokkaConfiguration.DokkaSourceSet>> {
+
+        fun getDependencies(sourceSet: DokkaConfiguration.DokkaSourceSet) : List<DokkaConfiguration.DokkaSourceSet> =
+            listOf(sourceSet) + configuration.sourceSets.filter {
+                it.sourceSetID in sourceSet.dependentSourceSets
+            }.flatMap { getDependencies(it) }
+
+        return configuration.sourceSets.map { it to getDependencies(it) }.toMap()
+}
```

Comment:
I don't know if this function should be here, or somewhere inside Dokka Configuration as it might be useful function in some cases. It gives you a map with dependencies (direct and transitive).